crowbar_framework/lib/crowbar/upgrade_status.rb

```diff
@@ -92,17 +92,22 @@ def start_step(step_name, out_of_order = false)
       end
     end
 
-    def end_step(success = true, errors = {})
+    def end_step(step_name = nil, out_of_order = false, success = true, errors = {})
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        unless running?
+        step_name ||= current_step
+        unless running? step_name
           @logger.warn("The step is not running, could not be finished")
           return false
         end
-        progress[:steps][current_step] = {
+        if step_name != current_step && !out_of_order
```

Comment:
> How can this ever be true

Easily, when user requests to start 3rd step while 2nd hasn't been run yet. But if user requests this with `out_of_order = true`, than it would be allowed.